src/org/openlcb/cdi/swing/CdiPanel.java

```diff
@@ -512,47 +554,145 @@ private void displayComplete() {
         }
         notifyTabColorRefresh();
         SwingUtilities.invokeLater(() -> {
-            Window win = SwingUtilities.getWindowAncestor(this);
-            if (win == null) {
+            JFrame f = new JFrame();
+            f = (JFrame)SwingUtilities.getAncestorOfClass(f.getClass(), this);
+            if (f == null) {
                 logger.log(Level.FINE, "Could not add close window listener");
                 return;
             }
-            win.addWindowListener(new WindowListener() {
+            f.setDefaultCloseOperation(WindowConstants.DO_NOTHING_ON_CLOSE);
+            f.addWindowListener(new WindowAdapter() {
                 @Override
-                public void windowOpened(WindowEvent windowEvent) {
+                public void windowClosing(WindowEvent e) {
+                    targetWindowClosingEvent(e);
                 }
+            });
+            _changeMade = false;
+            _saveButton.setEnabled(false);
+        });
+    }
 
-                @Override
-                public void windowClosing(WindowEvent windowEvent) {
-                    release();
-                    runUpdateComplete();
-                }
+    private void targetWindowClosingEvent(WindowEvent e) {
+        StringBuilder sb = new StringBuilder();
+        if (_saveNeeded) {
+            sb.append("The configuration was restored but not saved.");
+            sb.append("\n");
+        }
+        boolean save = _saveNeeded;
+        for (EntryPane entry : allEntries) {
+            if (entry.isDirty()) {
+                GetEntryNameVisitor nameGetter= new GetEntryNameVisitor(entry);
+                rep.visit(nameGetter);
+                sb.append(nameGetter.getName());
+                sb.append(" has not been saved.");
+                sb.append("\n");
+                save = true;
+            }
+        }
+        if (_panelChange) {
+            sb.append("The panel tables have been changed. To keep these changes, save the panel file.");
+            sb.append("\n");
+        }
+        if (sb.length() > 0) {
+            sb.append("\n");
+            if (save) {
+                sb.append("Do you want to save these changes?");
+            } else {
+                sb.append("Remember to save the panel file.");
+            }
+            int confirm = JOptionPane.showConfirmDialog(this, sb.toString(),
+                    "Save Changes", JOptionPane.YES_NO_OPTION,
+                    JOptionPane.QUESTION_MESSAGE);
+            if (confirm == JOptionPane.YES_OPTION && save) {
```

Comment:
This does not seem to actually save the changes. That's fine, but we should adjust the text to not promise the user that if they click the Yes button then their changes will be saved.

We could invert the message to the user:
"Warning: you have unsaved changes:
...
Would you like to discard these changes?"
Then we can have [Discard changes]  [Cancel] buttons or [Yes] [No] buttons. The first is a bit more clear to the user.
Upon clicking discard changes we proceed to close the window.

Reply:
It returns the user to the panel to selectively pick which changes to write to the node. Otherwise to do the above, it should walk through each change.